### PR TITLE
Rewards calculation for tBTCv2

### PR DESCRIPTION
**Pull request by dimpar**
This PR is copied from the [keep-core repository](https://github.com/keep-network/keep-core). The original PR and reviews were done [here](https://github.com/keep-network/keep-core/pull/3346)

Here we calculate rewards for the beta tBTCv2 operators. The output rewards JSON file should be used to generate a Merkle distribution tree.

There are 5 requirements for an operator to receive rewards in a given rewards interval:
- Random Beacon authorized
- tBTC authorized
- Total node(s) uptime 96% across all the instances after a peer joins the network. See examples
- Pre-generated params avg of 500 across all the instances. See examples
- Latest client build version. A 2-week update delay is allowed after the latest release.

E.g. 1
```
<operator_address>:
  "instances":
        "10.102.1.33:9701": {
          "upTimePercent": 50,
          "preParams": 400,
          "version": "v2.0.0-m1-9-g4049dc015"
        }
        "10.102.2.20:9601": {
          "upTimePercent": 46,
          "preParams": 600,
          "version": "v2.0.0-m1-9-g4049dc015"
        }
}
"upTimePercent" - total: 96% meets a 96% req across all the instances
"preParams" - total avg. 500 meets a 500 avg across all the instances
```

E.g. 2
```
<operator_address>:
  "instances":
        "10.102.1.33:9701": {
          "upTimePercent": 50,
          "preParams": 399,
          "version": "v2.0.0-m1-9-g4049dc015"
        }
        "10.102.2.20:9601": {
          "upTimePercent": 45,
          "preParams": 600,
          "version": "v2.0.0-m1-9-g4049dc015"
        }
}
"upTimePercent" - total: 95% does not meet a 96% req across all the instances
"preParams" - total avg. 499.5 does not meet a 500 avg across all the instances
```

`rewards.ts` will:
- call beacon and wallet registry contracts for authorization data. It calculates "weighted" authorization based on increase and decreases events during the interval. E.g.
```
- rewards interval Sep 1 - Sep 30
- beacon and tBTC authorization of 100k happened on Sep 1
- beacon and tBTC authorization increase event happened on Sep 10. It was increased to 150k

uptimeCoeficient: off-chain client's uptime
beaconWeightedAuthorization = 100k * 10/30 + 150k * 20/30
tbtcWeightedAuthorization = 100k * 10/30 + 150k * 20/30
weightedAuthorization = uptimeCoeficient * min(beaconWeightedAuthorization, tbtcWeightedAuthorization)
```
- query [Prometheus service](https://monitoring.test.threshold.network/prometheus/graph) for bootstrap peers' info
- aggregate and validate rewards requirements for each operator
- save rewards data for eligible operators

<details>
  <summary>Example operator details to calculate final rewards</summary>

```
   "0x5cD847903Bb7F29de77Eecc135628cA5b104A355": {
      "applications": {
        "beaconAuthorization": "800000000000000000000000",
        "tbtcAuthorization": "800000000000000000000000"
      },
      "instances": {
        "10.102.1.19:9601": {
          "upTimePercent": 74.52969075397857,
          "avgPreParams": 922,
          "version": "v2.0.0-m1-9-g4049dc015"
        },
        "10.102.1.77:9601": {
          "upTimePercent": 24.81376522052997,
          "avgPreParams": 971,
          "version": "v2.0.0-m1-9-g4049dc015"
        }
      },
      "requirements": {
        "isBeaconAuthorized": true,
        "isTbtcAuthorized": true,
        "isUptimeSatisfied": true,
        "isPreParamsSatisfied": true,
        "isVersionSatisfied": true
      }
    }
```

</details>

If all of the requirements are satisfied, then the rewards can be calculated with the following formula: 
```
amount = uptimeCoeficient * min(beaconWeightedAuthorization, tbtcWeightedAuthorization) * APR/12
```
APR = 15%

Example output format that is used to generate a merkle distribution tree. A parent key is a staking provider address, not operator. 

```
  {
    "0x0eC14BC7cCA82c942Cf276F6BbD0413216dDB2bE": {
      "beneficiary": "0x0eC14BC7cCA82c942Cf276F6BbD0413216dDB2bE",
      "amount": "9900000000000000000000"
    }
  },
```

### How to run
```
./rewards.sh --rewards-start-date Y-m-d --rewards-end-date Y-m-d --etherscan-token "<token>"
```

For more details that include optional flags please run `./rewards.sh --help`

### CODE REVIEW

File: src/tbtcv2-rewards/rewards.ts [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+    end: endRewardsTimestamp,
+    step: QUERY_STEP,
+  };
+
+  const bootstrapData = (
+    await queryPrometheus(queryBootstrapData, paramsBootstrapData)
+  ).data.result;
+
+  const operatorsData = new Array();
+  const rewardsData = new Array();
```
Comment (manumonti) [inline src/tbtcv2-rewards/rewards.ts:116]:
```suggestion
  const rewardsData: any = {};
```

File: src/tbtcv2-rewards/rewards.ts [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+    currentBlockNumber
+  );
+
+  for (let i = 0; i < bootstrapData.length; i++) {
+    const operatorAddress = bootstrapData[i].metric.chain_address;
+    let authorizations = new Map<string, BigNumber>(); // application: value
+    let requirements = new Map<string, boolean>(); // factor: true | false
+    let instancesData = new Map<string, Map<string, string | number>>();
+    let operatorData: any = {};
+    let rewardData: any = {};
```
Comment (manumonti) [inline src/tbtcv2-rewards/rewards.ts:174]:
```suggestion

```

File: src/tbtcv2-rewards/rewards.ts [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+      requirements.get(IS_VERSION_SATISFIED)
+    ) {
+      const beacon = BigNumber.from(authorizations.get(BEACON_AUTHORIZATION));
+      const tbct = BigNumber.from(authorizations.get(TBTC_AUTHORIZATION));
+      let minApplicationAuthorization = beacon;
+      if (beacon.gt(tbct)) {
+        minApplicationAuthorization = tbct;
+      }
+
+      rewardData[stakingProvider] = {
```
Comment (manumonti) [inline src/tbtcv2-rewards/rewards.ts:387]:
```suggestion
      rewardsData[stakingProvider] = {
```

File: src/tbtcv2-rewards/rewards.ts [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+        // amount = APR/12 * clientUptimeCoefficient * min(beaconWeightedAuthorization, tbtcWeightedAuthorization)
+        amount: minApplicationAuthorization
+          .mul(uptimeCoefficient)
+          .mul(monthlyRate)
+          .div(HUNDRED) // coefficient was multiplied by 100 earlier
+          .div(HUNDRED) // APR monthly rate was multiplied by 100 earlier
+          .div(HUNDRED) // APR monthly rate is in %
+          .toString(),
+      };
+      rewardsData.push(rewardData);
```
Comment (manumonti) [inline src/tbtcv2-rewards/rewards.ts:398]:
```suggestion

```

File: src/tbtcv2-rewards/rewards.ts [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+    }
+
+    operatorsData.push(operatorData);
+  }
+
+  console.log("operatorsData: ", JSON.stringify(operatorsData, null, 2));
+  console.log("rewardsData: ", JSON.stringify(rewardsData, null, 2));
+  fs.writeFileSync(rewardsDataOutput, JSON.stringify(rewardsData, null, 2));
```
Comment (manumonti) [inline src/tbtcv2-rewards/rewards.ts:406]:
```suggestion
  fs.writeFileSync(rewardsDataOutput, JSON.stringify(rewardsData, null, 4));
```

**Review by manumonti — COMMENTED** [review]
In order to have the maximum similarity between tBTCv2 rewards JSON and BonusRewards and OngoingRewards JSONs (https://github.com/threshold-network/merkle-distribution/blob/main/distributions/2022-09-30/MerkleInputOngoingRewards.json), can we have these changes?

File: src/tbtcv2-rewards/rewards.ts (hunk 1 above)
Comment (dimpar) [reply src/tbtcv2-rewards/rewards.ts]: Sure, part of [c78a6b7](https://github.com/threshold-network/merkle-distribution/pull/35/commits/c78a6b730f5fc7045bf5fbbd34b65be39292a8f4)

File: src/tbtcv2-rewards/rewards.ts (hunk 2 above)
Comment (dimpar) [reply src/tbtcv2-rewards/rewards.ts]: Sure, part of [c78a6b7](https://github.com/threshold-network/merkle-distribution/pull/35/commits/c78a6b730f5fc7045bf5fbbd34b65be39292a8f4)

File: src/tbtcv2-rewards/rewards.ts (hunk 3 above)
Comment (dimpar) [reply src/tbtcv2-rewards/rewards.ts]: Sure, part of [c78a6b7](https://github.com/threshold-network/merkle-distribution/pull/35/commits/c78a6b730f5fc7045bf5fbbd34b65be39292a8f4)

File: src/tbtcv2-rewards/rewards.ts (hunk 4 above)
Comment (dimpar) [reply src/tbtcv2-rewards/rewards.ts]: Sure, part of [c78a6b7](https://github.com/threshold-network/merkle-distribution/pull/35/commits/c78a6b730f5fc7045bf5fbbd34b65be39292a8f4)

File: src/tbtcv2-rewards/rewards.ts (hunk 5 above)
Comment (dimpar) [reply src/tbtcv2-rewards/rewards.ts:406]: Sure, part of [c78a6b7](https://github.com/threshold-network/merkle-distribution/pull/35/commits/c78a6b730f5fc7045bf5fbbd34b65be39292a8f4)

**Review by manumonti — APPROVED** [review]
LGTM ✌️

**Pull request merged**